### DISCUSSION TITLE
Just some questions

### DISCUSSION THREAD

**Discussion by SebastianStehle** [yantrajs/yantra#53] [General]
Hi,

I found YantraJS because I am missing some features in jint for many years now. I have some questions where I could not find answers for. My use case is a server application with many "adhoc" scripts.

1. Can you use asynchronous methods without the promises in javascript? I mean something that is actually async, but looks synchronous in javascript.
2. You mentioned that javascript is compiled to assemblies, but how will you remove them if they are not needed anymore?
3. Why do you need a synchronization context for async code? There is none in ASP.NET Core (afaik)
4. Do you have benchmarks to compare Jint to Yantra?
5. Do you have an example for Null coalesce?

**Comment by ackava** [comment]
> Can you use asynchronous methods without the promises in javascript? I mean something that is actually async, but looks synchronous in javascript.

I didn't get this, however `Task<T>` is safely translated to `Promise` and `Promise` can be translated to `Task<T>`.
Here are some of the samples, where you can find how easily you find how async code can be used in JavaScript as an async.
https://github.com/yantrajs/yantra/tree/main/YantraTests/Modules/in-built

In order to make async synchronous in JavaScript, you can create your own C# method, and use `AsyncPump.Run(() => SomeAsyncMethod())` and use it in JavaScript. This is how, `require` is supported in YantraJS along with `import`.

> You mentioned that javascript is compiled to assemblies, but how will you remove them if they are not needed anymore?

Currently they are compiled in memory, actually caching requires some extra code, so .NET CLR will remove them if you don't hold reference to the compiled script.

> Why do you need a synchronization context for async code? There is none in ASP.NET Core (afaik).

Promises are queued on to synchronization context, and Yantra doesn't know where it is running, whether it is ASP.NET Core or UI application. You can do `setTimeout` and `setInterval`, they need Synchronization Context, since JavaScript was originally designed to run as a single thread.

> Do you have benchmarks to compare Jint to Yantra?

Compared to JInt, Yantra actually generates IL code and executes it, so it will definitely be faster, but I haven't created any benchmarks.

> Do you have an example for Null coalesce?

Here are unit tests, where you will find example of each features, that are successfully executed in YantraJS.
https://github.com/yantrajs/yantra/tree/main/YantraJS.Core.Tests/Generator/Files/es6/Syntax
https://github.com/yantrajs/yantra/tree/main/YantraJS.Core.Tests/Generator/Files/es6/Async
https://github.com/yantrajs/yantra/tree/main/YantraJS.Core.Tests/Generator/Files/es6/Modules/clr

Have a look at this repository, 
https://github.com/yantrajs/web-site
In Website, we are using JavaScript as View in ASP.NET Core instead of CSHTML and we are also injecting dependencies in the View. It requires very little glue code to make it running.

And V8 Debugger integration is built on a different branch, we will soon make it live.

**Reply by SebastianStehle** [reply]
Thanks for your fast response.

I use Jint for squidex (https://github.com/squidex/squidex), where users can write scripts to react to events and build custom workflows,

Sometimes they want to call external APIs, so I have a helper methods for that. But jint does not support async/await yet and I do not want to block the thread so it looks like this currently:

```
getJSON(url, function (response) {
  ...
}); 
```

So if I understand you, I can implement a promise based function now, thne the script would look like this
```
const response = await getJSON(...);
```

But what I would like to achieve is this

```
const response = getJSON();
```

Code that looks like synchronously calls, but are actually asynchronous.

**Reply by ackava** [reply]
In your `getJSON`'s corresponding C# method, you can use `AsyncPump` as follow...

```C#

    context["getJSON"] = new JSFunction((in Arguments a) => {
          var url = a[0]?.ToString() ?? throw context.NewTypeError("url parameter is required");
          var response = AsyncPump.Run(() => httpClient.GetStringAsync(url));
          return new JSString(response);
    });

```

This way, your asynchronous C# function becomes synchronous JavaScript function.

However, I would still recommend supporting `await` in your code instead of turning async into sync.

YantraJS supports top level await as well. So if you create proper module system, and call your JavaScript from modules, it will run smoothly.

**Reply by SebastianStehle** [reply]
But would this not block the task and therefore the thread from the thread pool? I am not sure if I understand your AsyncPump properly.

**Reply by ackava** [reply]
Yes, it would block the current thread, but it will not crash or hang current thread as `AsyncPump` uses a `BlockingCollection` which will process all tasks as they are queued in `BlockingCollection`. This is where SynchronizationContext plays an important role. `AsyncPump` is a message processing logic used by `System.Threading.Tasks` namespace, so all of usual UI code will work correctly on it while awaiting the current operation.

I know that you can suspend execution in `JInt`, but you cannot suspend execution of JavaScript in YantraJS as the code is compiled.

However, if you use modules and if you make your code asynchronous, it would not block the current thread.

**Reply by ackava** [reply]
@SebastianStehle I request you to mark the answer to close this discussion.
